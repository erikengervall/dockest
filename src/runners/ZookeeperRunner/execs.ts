import execa from 'execa'

import { defaultDockerComposeRunOpts } from '../../constants'
import { DockestError } from '../../errors'
import { RunnerLogger } from '../../loggers'
import { acquireConnection, getContainerId, sleep, teardownSingle } from '../utils'
import { ZookeeperRunnerConfig } from './index'

interface Exec {
  start: (runnerConfig: ZookeeperRunnerConfig, runnerKey: string) => Promise<string>
  checkHealth: (runnerConfig: ZookeeperRunnerConfig, runnerKey: string) => Promise<void>
  teardown: (containerId: string, runnerKey: string) => Promise<void>
}

class ZookeeperExec implements Exec {
  private static instance: ZookeeperExec

  constructor() {
    return ZookeeperExec.instance || (ZookeeperExec.instance = this)
  }

  public start = async (runnerConfig: ZookeeperRunnerConfig, runnerKey: string) => {
    RunnerLogger.startContainer(runnerKey)

    const { port, service } = runnerConfig

    let containerId = await getContainerId(service)
    if (!containerId) {
      const portMapping = `--publish ${port}:2181`
      const cmd = `docker-compose run \
                    ${defaultDockerComposeRunOpts} \
                    ${portMapping} \
                    ${service}`
      RunnerLogger.shellCmd(cmd)
      await execa.shell(cmd)
    }
    containerId = await getContainerId(service)

    RunnerLogger.startContainerSuccess(service)

    return containerId
  }

  public checkHealth = async (runnerConfig: ZookeeperRunnerConfig, runnerKey: string) => {
    RunnerLogger.checkHealth(runnerKey)

    await this.checkConnection(runnerConfig, runnerKey)

    RunnerLogger.checkHealthSuccess(runnerKey)
  }

  public teardown = async (containerId: string, runnerKey: string) =>
    teardownSingle(containerId, runnerKey)

  private checkConnection = async (runnerConfig: ZookeeperRunnerConfig, runnerKey: string) => {
    const { connectionTimeout, port } = runnerConfig

    const recurse = async (connectionTimeout: number) => {
      RunnerLogger.checkConnection(runnerKey, connectionTimeout)

      if (connectionTimeout <= 0) {
        throw new DockestError('Zookeeper connection timed out')
      }

      try {
        await acquireConnection(port)

        RunnerLogger.checkConnectionSuccess(runnerKey)
      } catch (error) {
        connectionTimeout--

        await sleep(1000)
        await recurse(connectionTimeout)
      }
    }

    await recurse(connectionTimeout)
  }
}

export default ZookeeperExec
