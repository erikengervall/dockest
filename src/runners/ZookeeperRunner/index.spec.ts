import ZookeeperRunner from './index'

const config = {
  service: '_',
}
const ZookeeperRunner1 = new ZookeeperRunner(config)
const ZookeeperRunner2 = new ZookeeperRunner(config)

describe('ZookeeperRunner', () => {
  it('should create unique instances', () => {
    expect(ZookeeperRunner1).not.toBe(ZookeeperRunner2)
  })
})
