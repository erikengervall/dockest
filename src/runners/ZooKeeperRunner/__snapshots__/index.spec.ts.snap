// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZooKeeperRunner should create unique instances 1`] = `
ZooKeeperRunner {
  "containerId": "",
  "getComposeService": [Function],
  "initializer": "",
  "logger": Logger {
    "debug": [Function],
    "error": [Function],
    "info": [Function],
    "runnerService": "zk1",
    "runnerSymbol": "ðŸ¦‡ ",
    "setRunnerSymbol": [Function],
    "warn": [Function],
  },
  "runnerConfig": Object {
    "build": "",
    "commands": Array [],
    "connectionTimeout": 5,
    "dependsOn": Array [],
    "host": "localhost",
    "image": "some/image:123",
    "networks": undefined,
    "ports": Object {
      "2181": "2181",
    },
    "props": Object {},
    "responsivenessTimeout": 30,
    "service": "zk1",
  },
}
`;

exports[`ZooKeeperRunner should create unique instances 2`] = `
ZooKeeperRunner {
  "containerId": "",
  "getComposeService": [Function],
  "initializer": "",
  "logger": Logger {
    "debug": [Function],
    "error": [Function],
    "info": [Function],
    "runnerService": "zk2",
    "runnerSymbol": "ðŸ¦‡ ",
    "setRunnerSymbol": [Function],
    "warn": [Function],
  },
  "runnerConfig": Object {
    "build": "",
    "commands": Array [],
    "connectionTimeout": 5,
    "dependsOn": Array [],
    "host": "localhost",
    "image": "some/image:123",
    "networks": undefined,
    "ports": Object {
      "2181": "2181",
    },
    "props": Object {},
    "responsivenessTimeout": 30,
    "service": "zk2",
  },
}
`;
